File: templates/default/.agents/hooks/before_message.py
#!/usr/bin/env python3

"""Chat-local hook entry.

平台会在普通消息发给 opencode 之前执行这个脚本。

stdin 会收到一段 JSON，例如：
 {
   "provider": "feishu",
   "conversationId": "...",
   "conversationType": "group",
   "messageType": "text",
   "messageId": "om_xxx",
   "rootMessageId": "om_root_xxx",
   "parentMessageId": "om_parent_xxx",
   "threadId": "omt_xxx",
   "senderType": "user",
   "senderId": "ou_xxx",
   "text": "用户原始文本",
   "attachments": [{"mime": "image/png", "filename": "a.png", "url": "file:///..."}]
  }

stdout 留空表示不改动，继续按原消息处理。

stdout 也可以输出 JSON：
- {"drop": true}
  直接拦截，不再进入 opencode。
- {"drop": true, "reactionEmoji": "ENOUGH"}
  直接拦截，并给这条消息补一个 reaction。
- {"drop": true, "replyText": "..."}
  拦截，并回一条文本给当前聊天。
- {"text": "改写后的文本"}
  改写发给 opencode 的文本内容。
- {"text": "改写后的文本", "systemText": "给模型的 system 提示"}
  改写发给 opencode 的文本内容，并额外指定这条消息的 system 提示。

当前只支持改写 text，不支持改 attachments。

如果 messageType 是 `interactive`，那么 `text` 里会是卡片消息的原始 content JSON。

如果当前消息是在某个 topic / thread 下回复，`rootMessageId` / `parentMessageId` /
`threadId` 会把回复上下文一起带进来，hook 可以自己据此关联原始消息。
"""


def main() -> int:
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: internal/flow/before_message.go
package flow

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/chenxuan520/agentbot/internal/backend"
	"github.com/chenxuan520/agentbot/internal/conversation"
)

const beforeMessageHookRelativePath = ".agents/hooks/before_message.py"

type beforeMessageHookRunner func(context.Context, string, TextInput) (beforeMessageHookResult, error)

type beforeMessageHookResult struct {
	Drop          bool   `json:"drop"`
	ReplyText     string `json:"replyText"`
	Text          string `json:"text"`
	SystemText    string `json:"systemText"`
	ReactionEmoji string `json:"reactionEmoji"`
}

func runBeforeMessageHook(ctx context.Context, workspacePath string, input TextInput) (beforeMessageHookResult, error) {
	hookPath := filepath.Join(workspacePath, beforeMessageHookRelativePath)
	info, err := os.Stat(hookPath)
	if err != nil {
		if os.IsNotExist(err) {
			return beforeMessageHookResult{}, nil
		}
		return beforeMessageHookResult{}, err
	}
	if info.IsDir() {
		return beforeMessageHookResult{}, fmt.Errorf("before_message hook is a directory: %s", hookPath)
	}

	payload := buildBeforeMessageHookPayload(input)
	data, err := json.Marshal(payload)
	if err != nil {
		return beforeMessageHookResult{}, err
	}

	cmd := exec.CommandContext(ctx, "python3", hookPath)
	cmd.Dir = workspacePath
	cmd.Stdin = bytes.NewReader(data)
	cmd.Env = buildBeforeMessageHookEnv(workspacePath, input)

	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		detail := strings.TrimSpace(stderr.String())
		if detail == "" {
			detail = err.Error()
		}
		return beforeMessageHookResult{}, fmt.Errorf("run before_message.py: %s", detail)
	}

	raw := strings.TrimSpace(stdout.String())
	if raw == "" {
		return beforeMessageHookResult{}, nil
	}

	var result beforeMessageHookResult
	if err := json.Unmarshal([]byte(raw), &result); err != nil {
		return beforeMessageHookResult{}, fmt.Errorf("decode before_message.py output: %w", err)
	}
	return result, nil
}

func buildBeforeMessageHookPayload(input TextInput) map[string]any {
	legacyType := legacyHookConversationType(input.ConversationType)
	return map[string]any{
		"provider":         input.Provider,
		"conversationId":   input.ConversationID,
		"conversationType": input.ConversationType,
		"chatType":         legacyType,
		"messageType":      input.MessageType,
		"messageId":        input.MessageID,
		"rootMessageId":    input.RootMessageID,
		"parentMessageId":  input.ParentMessageID,
		"threadId":         input.ThreadID,
		"senderType":       input.SenderType,
		"senderId":         input.SenderID,
		"senderOpenId":     input.SenderID,
		"eventAction":      input.EventAction,
		"reactionEmoji":    input.ReactionEmoji,
		"text":             input.Text,
		"attachments":      marshalHookAttachments(input.Attachments),
	}
}

func buildBeforeMessageHookEnv(workspacePath string, input TextInput) []string {
	legacyType := legacyHookConversationType(input.ConversationType)
	return append(os.Environ(),
		"AGENT_BOT_WORKSPACE="+workspacePath,
		"AGENT_BOT_PROVIDER="+input.Provider,
		"AGENT_BOT_CONVERSATION_ID="+input.ConversationID,
		"AGENT_BOT_CONVERSATION_TYPE="+input.ConversationType,
		"AGENT_BOT_CHAT_TYPE="+legacyType,
		"AGENT_BOT_MESSAGE_TYPE="+input.MessageType,
		"AGENT_BOT_MESSAGE_ID="+input.MessageID,
		"AGENT_BOT_ROOT_MESSAGE_ID="+input.RootMessageID,
		"AGENT_BOT_PARENT_MESSAGE_ID="+input.ParentMessageID,
		"AGENT_BOT_THREAD_ID="+input.ThreadID,
		"AGENT_BOT_SENDER_TYPE="+input.SenderType,
		"AGENT_BOT_SENDER_ID="+input.SenderID,
		"AGENT_BOT_SENDER_OPEN_ID="+input.SenderID,
		"AGENT_BOT_EVENT_ACTION="+input.EventAction,
		"AGENT_BOT_REACTION_EMOJI="+input.ReactionEmoji,
	)
}

func legacyHookConversationType(conversationType string) string {
	switch conversation.NormalizeType(conversationType) {
	case conversation.TypeGroup:
		return "group"
	case conversation.TypeThread:
		return "topic_group"
	default:
		return "p2p"
	}
}

func marshalHookAttachments(attachments []backend.Attachment) []map[string]string {
	if len(attachments) == 0 {
		return nil
	}
	result := make([]map[string]string, 0, len(attachments))
	for _, attachment := range attachments {
		result = append(result, map[string]string{
			"mime":     attachment.Mime,
			"filename": attachment.Filename,
			"url":      attachment.URL,
		})
	}
	return result
}
